File: document/kernel/support.js
'use strict';
globalThis.kernel.support=(()=>{
    function consequence(rule){
        const remaining=rule.map(value=>value.premise.length),dependency=new Map(),answer=new Set(),pending=[];
        rule.forEach((value,index)=>{
            if(!value.premise.length)pending.push(value.head);
            for(const atom of value.premise)dependency.set(atom,[...(dependency.get(atom)||[]),index]);
        });
        for(let position=0;position<pending.length;position++){
            const atom=pending[position];if(answer.has(atom))continue;answer.add(atom);
            for(const index of dependency.get(atom)||[]){remaining[index]--;if(!remaining[index])pending.push(rule[index].head);}
        }
        return answer;
    }
    function evaluate(model){
        const answer=consequence(model.clause),atom=new Set(model.clause.flatMap(value=>[value.head,...value.premise]));
        return {status:atom=>answer.has(atom)?'supported':'unsupported',supported:[...answer].sort(),unsupported:[...atom].filter(value=>!answer.has(value)).sort(),closed:model.closed};
    }
    return {consequence,evaluate};
})();

File: system/support.rs
use serde::Serialize;
use std::collections::{BTreeSet, HashMap, HashSet, VecDeque};

#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub enum Atom {
    State(usize),
    Event(usize),
    View(usize),
}

#[derive(Clone, Debug, Eq, Hash, PartialEq)]
pub struct Clause {
    pub head: Atom,
    pub premise: BTreeSet<Atom>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum Status {
    Supported,
    Unsupported,
}

pub struct Support {
    established: HashSet<Atom>,
}

impl Support {
    pub fn new(clause: impl IntoIterator<Item = Clause>) -> Self {
        let clause = clause.into_iter().collect::<Vec<_>>();
        let mut remaining = clause
            .iter()
            .map(|clause| clause.premise.len())
            .collect::<Vec<_>>();
        let mut dependency = HashMap::<Atom, Vec<usize>>::new();
        let mut pending = VecDeque::new();
        for (index, clause) in clause.iter().enumerate() {
            if clause.premise.is_empty() {
                pending.push_back(clause.head);
            }
            for &atom in &clause.premise {
                dependency.entry(atom).or_default().push(index);
            }
        }
        let mut established = HashSet::new();
        while let Some(atom) = pending.pop_front() {
            if !established.insert(atom) {
                continue;
            }
            for &index in dependency.get(&atom).into_iter().flatten() {
                remaining[index] -= 1;
                if remaining[index] == 0 {
                    pending.push_back(clause[index].head);
                }
            }
        }
        Self { established }
    }

    pub fn status(&self, atom: Atom) -> Status {
        if self.established.contains(&atom) {
            Status::Supported
        } else {
            Status::Unsupported
        }
    }
}
